### DISCUSSION TITLE
Test in Docker container fails with error 137 indicating a probable memory leak

### DISCUSSION THREAD

**Discussion by vishucrafts** [community/community#25492] [Actions]
<pre><code class="lang-auto">name: Docker Image CI

on: [push]

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v1
      - name: Build the docker-compose stack
        run: docker-compose -f docker-compose.yaml up -d
      - name: Check running containers
        run: docker ps -a
      - name: Check logs
        run: docker logs web_container
      - name: Run jest test suite
        run: docker exec -w /home/node/app web_container npm run test
      - name: Run cypress test suite
        run: docker exec -w /home/node/app web_container npm run test:e2e
</code></pre>
<p>This command is supposed to run jest</p>
<pre><code class="lang-auto">docker exec -w /home/node/app web_container npm run test
</code></pre>
<p>and this is supposed to run cypress</p>
<pre><code class="lang-auto">docker exec -w /home/node/app web_container npm run test
</code></pre>
<p>My repository works fine in local computer but fails in github actions.</p>
<p>Link to my repository: <a href="https://github.com/pvgdev/nextjs-template" rel="noopener nofollow ugc">https://github.com/pvgdev/nextjs-template</a></p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/pvgdev">@pvgdev</a>,</p>
<p>Error 137 in Docker denotes that the container was killed by ‘OOM killer’ (Out of Memory). This happens when there isn’t enough memory in the container for running the process.<br>
The 2 main out-of-memory reasons:</p>
<ol>
<li>
<p><strong>Docker container has run out of memory</strong><br>
By default, Docker containers use the available memory in the host machine. To prevent a single container from abusing the host resources, there may be some memory limits set for per container on the virtual machines of GitHub-hosted runners.<br>
It is important not to allow a running container to consume too much of the host machine’s memory.</p>
</li>
<li>
<p><strong>Docker host has no free memory</strong><br>
Each virtual machine of GitHub-hosted runner has the same hardware resources available.</p>
<ul>
<li>2-core CPU</li>
<li>7 GB of RAM memory</li>
<li>14 GB of SSD disk space</li>
</ul>
<p>The memory limit that can be allotted to the Docker containers is limited by the total available memory in the host machine which hosts them.<br>
Many often, when usage and traffic increases, the available free memory may be insufficient for all the containers. As a result, containers may crash.</p>
</li>
</ol>
<p>To resolve error 137 in Docker, you can try with the following steps:</p>
<ol>
<li>
<p><strong>Optimize the services</strong><br>
Unoptimized applications could take up more than optimal memory. So, the first step is to monitor the application running in the container and to optimize the service. This can be done by editing the configuration file or recompiling the service.</p>
</li>
<li>
<p><strong>Mount config files from outside</strong><br>
It is always advisable to mount the config files of services from outside the Docker container. This will allow to edit them easily without recompiling the Docker image.</p>
</li>
</ol>
<p>To view more details about container’s resource constraints, you can see “<a href="https://docs.docker.com/config/containers/resource_constraints/" rel="noopener nofollow ugc"><strong>Runtime options with Memory, CPUs, and GPUs</strong></a>”.</p>
